### PR TITLE
Performance enhancements for deep circuits

### PR DESCRIPTION
**Pull request by albi3ro**
Recent performance work has been looking at deep circuits created from decomposing `qml.QuantumPhaseEstimation`.

```
phase = 5
target_wires = [0,1,2]
#unitary = qml.RX(phase, wires=0)
unitary = qml.RX(phase, wires=[2]) @ qml.CNOT(wires=[0, 1])

n_estimation_wires = 14
estimation_wires = range(len(target_wires), n_estimation_wires + len(target_wires))

dev = qml.device('lightning.qubit', wires=n_estimation_wires + len(target_wires))

@qml.qnode(dev, diff_method=None)
def circuit():
    # Start in the |+> eigenstate of the unitary
    ops = []
    for i in target_wires:
        ops.append(qml.Hadamard(wires=i))

    ops.append(qml.QuantumPhaseEstimation(
        unitary,
        #target_wires=target_wires,
        estimation_wires=estimation_wires,
    ))

    return qml.expval(qml.PauliZ(0))
```

Profiling shows substantial python overheads. A local run with snakeviz on my computer shows:

* 88% of total time spent in `pennylane.tape.tape.expand_tape`
* 59% of total time spent in `pennylane.queuing.apply`
* 10% of total time spend in `QuantumScript.from_queue`

I see two takeaways from these numbers.

1) We are spending way too much time in internal queueing where we don't need to spend time
2) `apply` has something very problematic going on.

## Stopping using internal queueing

By avoiding using queuing internally on `Operator.expand`, we can cut off over half the time.

Screenshot 2023-07-12 at 4 59 57 PM

## Turn off `QuantumScript` initialization in `Operator.expand()`

Screenshot 2023-07-13 at 10 13 08 AM

During decomposition, a lot of time was being spent on initialization properties in `QuantumScript` that were never going to get used.  By being able to selectively turn off computing those properties, we were able to shave off another 1.5 seconds.

## Change to `qml.apply`

When we stop using queuing for the expansion, we no longer have the same problem with `qml.apply` in this particular example.

Unfortunately, the problematic scaling for `qml.apply` still exists and will cause problems in other circumstances.

To illustrate this example, we start with an `AnnotatedQueue` with 10, 000 entries, and we time adding one more operator with `qml.apply`.

```
op = qml.S(0)

with qml.queuing.AnnotatedQueue() as q:
    for _ in range(10000):
        qml.apply(op)
```

Screenshot 2023-07-12 at 5 03 35 PM

Basically, `qml.apply` was converting the entire queue into a list, then running comparison to see if the operator was in the list. Now we just check in the object is in the queue directly.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -35,6 +35,11 @@
 
 <h3>Breaking changes 💔</h3>
 
+* `Operator.expand` now relies directly on the output of `Operator.decomposition`, rather on the what it queues.
+  This has substantial performance improvements for deep circuits, particularily for decompositions that would
+  have to use `qml.apply`.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:40]:
```suggestion
  have to use `qml.apply`.
  [(#4355)](https://github.com/PennyLaneAI/pennylane/pull/4355)
```

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -33,8 +33,16 @@
 * The default label for a `StatePrep` operator is now `|Ψ⟩`.
   [(#4340)](https://github.com/PennyLaneAI/pennylane/pull/4340)
 
+* `Operator.expand` now has a keyword argument ``_update=True``. Turning this to ``False`` improves efficiency,
+  but some properties will not be precomputed on initialization.
+  [(#4355)](https://github.com/PennyLaneAI/pennylane/pull/4355) 
+
 <h3>Breaking changes 💔</h3>
 
+* `Operator.expand` now relies directly on the output of `Operator.decomposition`, rather than on the what it queues.
```
Comment (mlxd) [inline doc/releases/changelog-dev.md:42]:
```suggestion
* `Operator.expand` now relies directly on the output of `Operator.decomposition`, rather than on what it queues.
```

?

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -33,8 +33,16 @@
...
+  This change provides substantial performance improvements operators with large decompositions.
```
Comment (mlxd) [inline doc/releases/changelog-dev.md:43]:
```suggestion
  This change provides substantial performance improvements for operators with large decompositions.
```

File: pennylane/templates/subroutines/qsvt.py [hunk 4]
```diff
@@ -321,17 +321,20 @@ def compute_decomposition(
...
+                    qml.apply(UA)
                 op_list.append(UA)
 
             else:
                 op_list.append(adjoint(UA_adj))
 
-        qml.apply(projectors[-1])
+        if qml.QueuingManager.recording():
+            qml.apply(projectors[-1])
         op_list.append(projectors[-1])
```
Comment (mlxd) [inline pennylane/templates/subroutines/qsvt.py:338]: Will this need to become a rule of thumb for all template implementations?

**Review by mlxd — COMMENTED** [review]
Just a few small things and a quick question

File: pennylane/templates/subroutines/qsvt.py (hunk 4 above)
Comment (albi3ro) [reply pennylane/templates/subroutines/qsvt.py:338]: It already was.

We just didn't catch that it because there are no unit tests for `QSVT.compute_decomposition`.

File: pennylane/tape/tape.py [hunk 5]
```diff
@@ -201,7 +201,7 @@ def stop_at(obj):  # pylint: disable=unused-argument
             if isinstance(obj, (Operator, qml.measurements.MeasurementProcess)):
                 # Object is an operation; query it for its expansion
                 try:
-                    obj = obj.expand()
+                    obj = obj.expand(_update=False)
```
Comment (albi3ro) [inline pennylane/tape/tape.py:204]:
```suggestion
                    obj = obj.expand()
```

File: tests/test_operation.py [hunk 6]
```diff
@@ -273,6 +273,22 @@ def __init__(self, wires, basis_state=None):  # pylint:disable=super-init-not-ca
...
+        class CustomOp(Operator):
+            """Dummy operator."""
+
+            def decomposition(self):
+                return [qml.RX(1.2, wires=0)]
+
+        op = CustomOp(wires=0)
+        out = op.expand()
+
+        assert not out._par_info
```
Comment (timmysilv) [inline tests/test_operation.py:289]: could you add something like
```
out._update()
assert out._par_info == [{...}]

**Review by timmysilv — APPROVED** [review]
one tiny ask but otherwise this is huge!! very excited to get this merged (assuming the test failures are from explicitly testing something that this PR breaks, but if not I'll re-review again later)

File: pennylane/transforms/op_transforms.py [hunk 7]
```diff
@@ -261,16 +261,18 @@ def fn(self, obj, *args, **kwargs):
             try:
                 # attempt to decompose the operation and call
                 # the tape transform function if defined
+                print(obj)
+                print(obj.expand())
```
Comment (albi3ro) [inline pennylane/transforms/op_transforms.py:265]:
```suggestion
```

File: doc/releases/changelog-dev.md [hunk 8]
```diff
@@ -40,6 +40,10 @@
 * The experimental device interface is integrated with the `QNode` for Jax.
   [(#4323)](https://github.com/PennyLaneAI/pennylane/pull/4323)
 
+* `tape_expand` now uses `Operator.decomposition` instead of `Operator.expand` for in order to make
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:43]:
```suggestion
* `tape_expand` now uses `Operator.decomposition` instead of `Operator.expand` in order to make
```

File: tests/interfaces/test_autograd_qnode.py [hunk 9]
```diff
@@ -500,16 +499,13 @@ def test_differentiable_expand(self, interface, dev_name, diff_method, grad_on_e
...
-                with qml.queuing.AnnotatedQueue() as q_tape:
-                    qml.Rot(lam, theta, -lam, wires=wires)
-                    qml.PhaseShift(phi + lam, wires=wires)
-
-                tape = QuantumScript.from_queue(q_tape)
-                return tape
+                return [
+                    qml.Rot(lam, theta, -lam, wires=wires),
+                    qml.PhaseShift(phi + lam, wires=wires),
+                ]
```
Comment (vincentmr) [inline tests/interfaces/test_autograd_qnode.py:508]: It appears this is coming back a lot, should we define and import that from some other module?

**Review by vincentmr — APPROVED** [review]
I'll just make this [suggestion](https://github.com/PennyLaneAI/pennylane/pull/4355#discussion_r1267054179) for an optional change. Otherwise, happy to approve.

File: tests/interfaces/test_autograd_qnode.py (hunk 9 above)
Comment (albi3ro) [reply tests/interfaces/test_autograd_qnode.py:508]: So yes, we do need to test once for every interface, once for old default qubit and once for new default qubit, and once for new return types and once for old return types.

So that does add up into a lot of tests that look nearly identical.

But I'm fine with verbosity in tests, as they are often easier to debug when they fail.  And while I did have to make this change in a bunch of places, it was a fairly easy change to make.
